src/ptrace.cc

```diff
@@ -93,4 +150,30 @@ std::unique_ptr<uint8_t[]> PtracePeekBytes(pid_t pid, unsigned long addr,
   }
   return bytes;
 }
+
+#ifdef __amd64__
+long PtraceCallFunction(pid_t pid, unsigned long addr) {
```

Comment:
I think you're almost certainly right. In very near the limits of my understanding on this stuff - this code is crudely adapted from some code you wrote. I left out the mmap stuff mostly because this seemed to work, but I did find that if I called this on every profile cycle that sometimes the SIGTRAP would leak out, which this may explain.